### PR TITLE
[IOTDB-6121] Consensus layer interface and exception handling refactoring

### PR DESCRIPTION
**Pull request by OneSizeFitsQuorum**
see [jira](https://issues.apache.org/jira/browse/IOTDB-6121)

### CODE REVIEW

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java [hunk 1]
```diff
@@ -112,7 +112,8 @@ class RatisConsensus implements IConsensus {
   private final RaftProperties properties = new RaftProperties();
   private final RaftClientRpc clientRpc;
 
-  private final IClientManager<RaftGroup, RatisClient> clientManager;
+  private final IClientManager<RaftGroup, org.apache.iotdb.consensus.ratis.RatisClient>
```
Comment (SzyWilliam) [inline iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java:115]: Why use the full package name?

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java [hunk 2]
```diff
@@ -243,117 +243,121 @@ private RaftClientReply writeLocallyWithRetry(RaftClientRequest request) throws
...
-
     Message ret = reply.getMessage();
-    ResponseMessage readResponseMessage = (ResponseMessage) ret;
-    DataSet dataSet = (DataSet) readResponseMessage.getContentHolder();
-    return ConsensusReadResponse.newBuilder().setDataSet(dataSet).build();
+    org.apache.iotdb.consensus.ratis.ResponseMessage readResponseMessage =
+        (org.apache.iotdb.consensus.ratis.ResponseMessage) ret;
+    return Optional.ofNullable(readResponseMessage.getContentHolder())
+        .map(DataSet.class::cast)
+        .orElse(null);
```
Comment (SzyWilliam) [inline iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java:360]: If the returned read response is null, we should force callers to handle this situation.
Or we can simply assert it is non-null.

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java [hunk 3]
```diff
@@ -243,117 +243,121 @@ private RaftClientReply writeLocallyWithRetry(RaftClientRequest request) throws
...
+        RaftClientReply localServerReply = writeLocallyWithRetry(clientRequest);
         if (localServerReply.isSuccess()) {
-          ResponseMessage responseMessage = (ResponseMessage) localServerReply.getMessage();
-          TSStatus writeStatus = (TSStatus) responseMessage.getContentHolder();
-          return ConsensusWriteResponse.newBuilder().setStatus(writeStatus).build();
+          org.apache.iotdb.consensus.ratis.ResponseMessage responseMessage =
+              (org.apache.iotdb.consensus.ratis.ResponseMessage) localServerReply.getMessage();
+          return Optional.ofNullable(responseMessage.getContentHolder())
+              .map(TSStatus.class::cast)
+              .orElse(null);
```
Comment (SzyWilliam) [inline iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java:292]: The same.

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java [hunk 4]
```diff
@@ -243,117 +243,121 @@ private RaftClientReply writeLocallyWithRetry(RaftClientRequest request) throws
...
     if (suggestedLeader != null) {
       TEndPoint leaderEndPoint = Utils.fromRaftPeerAddressToTEndPoint(suggestedLeader.getAddress());
       writeResult.setRedirectNode(new TEndPoint(leaderEndPoint.getIp(), leaderEndPoint.getPort()));
     }
-    return ConsensusWriteResponse.newBuilder().setStatus(writeResult).build();
+    return writeResult;
   }
 
-  /** Read directly from LOCAL COPY notice: May read stale data (not linearizable) */
+  /** Read directly from LOCAL COPY notice. */
```
Comment (SzyWilliam) [inline iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java:324]: Current read implementation still may read stale data (not linearizable)

File: iotdb-core/consensus/src/test/java/org/apache/iotdb/consensus/ratis/RatisConsensusTest.java [hunk 5]
```diff
@@ -220,11 +220,15 @@ private void doConsensus(IConsensus consensus, ConsensusGroupId gid, int count,
           () -> {
             ByteBufferConsensusRequest incrReq = TestUtils.TestRequest.incrRequest();
 
-            ConsensusWriteResponse response = consensus.write(gid, incrReq);
-            if (response.getException() != null) {
-              response.getException().printStackTrace(System.out);
+            TSStatus response;
+            try {
+              response = consensus.write(gid, incrReq);
+            } catch (ConsensusException e) {
```
Comment (SzyWilliam) [inline iotdb-core/consensus/src/test/java/org/apache/iotdb/consensus/ratis/RatisConsensusTest.java:226]: Shall we fail the test immediately when an exception is thrown?

**Review by SzyWilliam — COMMENTED** [review]
Thanks @OneSizeFitsQuorum and @BUAAserein a lot for working on this! Quite a huge patch. Have a few issues we can discuss on.

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java (hunk 1 above)
Comment (OneSizeFitsQuorum) [reply iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java:116]: Fixed

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java (hunk 2 above)
Comment (OneSizeFitsQuorum) [reply iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java:361]: A null value here indicates a bug in the code, and avoiding an NPE isn't going to make much sense at this point, so I'll reverse the change

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java (hunk 3 above)
Comment (OneSizeFitsQuorum) [reply iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java:293]: Fixed

File: iotdb-core/consensus/src/test/java/org/apache/iotdb/consensus/ratis/RatisConsensusTest.java (hunk 5 above)
Comment (OneSizeFitsQuorum) [reply iotdb-core/consensus/src/test/java/org/apache/iotdb/consensus/ratis/RatisConsensusTest.java:254]: It will fail in line 259

**Review by BUAAserein — APPROVED** [review]
LGTM

**Review by CRZbulabula — CHANGES_REQUESTED** [review]
A few naming problems

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/IConsensus.java [hunk 6]
```diff
@@ -19,137 +19,179 @@
...
   /**
-   * Tell the group to update an active Peer. The modifiable part of {@link Peer} is TEndPoint.
+   * Transfer the leadership to other peer to meet some load balancing needs.
    *
-   * @param groupId the consensus group this peer belongs
-   * @param oldPeer the peer to be updated
-   * @param newPeer the peer to be updated to
+   * @param groupId the consensus group which should execute this command
+   * @param newLeader the target leader peer
+   * @throws ConsensusGroupAlreadyExistException when the specified consensus group doesn't exist
```
[the remark this answers is not in the document]
Comment (OneSizeFitsQuorum) [reply iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/IConsensus.java:153]: Fixed

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/IConsensus.java [hunk 7]
```diff
@@ -19,137 +19,179 @@
...
-   * delete local data and do other cleanup works. Be sure this method is called after successfully
-   * removing this peer from current consensus group configuration (by calling {@link
-   * #removePeer(ConsensusGroupId, Peer)} or {@link #changePeer(ConsensusGroupId, List)}).
+   * delete local data and do other cleanup works. deleteLocalPeer should be called after
+   * successfully removing this peer from current consensus group configuration (by calling {@link
+   * #removeRemotePeer(ConsensusGroupId, Peer)}).
    *
-   * @param groupId the consensus group this Peer used to belong
+   * @param groupId the consensus group this peer used to belong
+   * @throws ConsensusGroupAlreadyExistException when the specified consensus group doesn't exist
```
[the remark this answers is not in the document]
Comment (OneSizeFitsQuorum) [reply iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/IConsensus.java:109]: Fixed

**Review by CRZbulabula — APPROVED** [review]
LGTM!

**Pull request merged**